service/tablet_allocator.cc

```diff
@@ -342,12 +411,82 @@ class load_balancer {
         co_return std::move(plan);
     }
 
+    future<migration_plan> make_resize_plan() {
+        migration_plan plan;
+
+        if (!_tm->tablets().balancing_enabled()) {
+            co_return std::move(plan);
+        }
+
+        cluster_resize_stats resize_load;
+
+        for (auto&& [table, tmap_] : _tm->tablets().all_tables()) {
+            auto& tmap = tmap_;
+
+            auto avg_tablet_size = _stats.load_stats_for_table(table).size_in_bytes / std::max(tmap.tablet_count(), size_t(1));
+            resize_load.update(table, avg_tablet_size, tmap.resize());
+            lblogger.info("Table {} with tablet_count={} has an average tablet size of {}", table, tmap.tablet_count(), avg_tablet_size);
+            co_await coroutine::maybe_yield();
+        }
+
+        // Emit new resize decisions
+
+        table_resize_plan resize_plan;
+
+        auto resize_units = ssize_t(max_concurrent_resize_requests) - ssize_t(resize_load.tables_being_resized.size());
+        std::make_heap(resize_load.tables_need_resize.begin(), resize_load.tables_need_resize.end(), resize_load.resize_urgency_cmp());
+        while (resize_load.tables_need_resize.size() && resize_units-- > 0) {
+            auto& [table, avg_tablet_size] = resize_load.tables_need_resize.front();
+
+            if (!resize_load.table_needs_resize(avg_tablet_size)) {
+                // shouldn't happen, but skip it.
+                continue;
+            }
+
+            auto resize_decision = cluster_resize_stats::to_resize_decision(avg_tablet_size);
+            lblogger.info("Emitting resize decision of type {} for table {} due to avg tablet size of {}", resize_decision.resize_name(), table, avg_tablet_size);
+            resize_plan.resize[table] = std::move(resize_decision);
+            _stats.for_cluster().resizes_emitted++;
+
+            std::pop_heap(resize_load.tables_need_resize.begin(), resize_load.tables_need_resize.end(), resize_load.resize_urgency_cmp());
+            resize_load.tables_need_resize.pop_back();
+        }
+
+        // Revoke resize decision if any table no longer needs it
+        // Also communicate coordinator if any table is ready for finalizing resizing
+
+        for (auto& [table, avg_tablet_size] : resize_load.tables_being_resized) {
+            if (!resize_load.table_needs_resize(avg_tablet_size)) {
+                resize_plan.resize[table] = cluster_resize_stats::revoke_resize_decision();
+                _stats.for_cluster().resizes_revoked++;
+                lblogger.info("Revoking resize decision for table {} due to avg tablet size of {}", table, avg_tablet_size);
+                continue;
+            }
+
+            auto& tmap = _tm->tablets().get_tablet_map(table);
+            auto& tstats = _stats.load_stats_for_table(table);
+
+            // If all replicas mirrored the current phase, it means that this table is
+            // ready for finalizing the resizing request.
```

Comment:
We will need a way to handle dead nodes. We can just ignore them, but when they come back, we can't migrate any tablets off them until they complete the split.

But okay for now, this isn't urgent. Down nodes are rare, resize decisions are rare, rare * rare = rare^2.

Reply:
Theoretically, a tablet can be migrated concurrently to the split decision. We'd need to make sure that the source keeps the (non split ready state) until migration completes, such that topology change cannot progress. However, it could happen that the destination node is already seen as split ready by coordinator, so once tablet is migrated (without completing splitting), then we can get incorrect behavior. So migration has to wait for splitting completeness. Any alternative to that?

Reply:
It's likely* to happen: table grows quickly, topology coordinator decides to split, in parallel cluster manager decides to add a new node.

We collect completion by rpc? If so just return a list of tablet replicas with completion status. This way a migrating tablet can't escape.

Also, the coordinator knows it's migrating tablets and knows when migration completes, it can invalidate the completion state of a node if it received a new tablet from an incomplete node.


* actually not very likely